### PR TITLE
add code for detecting (and killing) a hung task manager task

### PR DESCRIPTION
**Pull request by ryanpetrello**
this is a hail-mary, worst-case scenario, save-the-AWX-install-from-being-totally-stuck sort of thing

see: https://github.com/ansible/awx/issues/5617

I tested this by intentionally making the task manager never yield:

```diff
diff --git a/awx/main/scheduler/task_manager.py b/awx/main/scheduler/task_manager.py
index 9cd0aa1db9..854dfbcf74 100644
--- a/awx/main/scheduler/task_manager.py
+++ b/awx/main/scheduler/task_manager.py
@@ -543,6 +543,11 @@ class TaskManager():
     def _schedule(self):
         finished_wfjs = []
         all_sorted_tasks = self.get_tasks()
+        while True:
+            import time
+            time.sleep(1)
+            # locked!
```

image

### CODE REVIEW

**Comment by ryanpetrello** [comment]
bonus:

image

File: awx/main/dispatch/pool.py [hunk 1]
```diff
@@ -365,6 +367,28 @@ def cleanup(self):
...
                 self.workers.remove(w)
+            if w.alive:
+                # if we discover a task manager invocation that's been running
+                # too long, reap it (because otherwise it'll just hold the postgres
+                # advisory lock forever); the goal of this code is to discover
+                # deadlocks or other serious issues in the task manager that cause
+                # the task manager to never do more work
+                current_task = w.current_task
+                if current_task:
+                    if current_task.get('task', '').endswith('tasks.run_task_manager'):
```
Comment (AlanCoding) [inline awx/main/dispatch/pool.py:378]: I get confused in the dispatcher code about whether these messages are always dicts, or sometimes strings.

Well if `current_task` is a string, I don't really care what it does. Just don't do anything would be fine.

**Comment by AlanCoding** [comment]
Some of these arguments (a transaction inside of a advisory lock), but not necessarily all (sending websockets as an on_commit action) could also apply to

https://github.com/ansible/awx/blob/4912cbd2dac7ad0ada10e353fb375f3607f85f97/awx/main/tasks.py#L155

There's no argument _against_ reaping this task after 5 minutes. Maybe same applies to some other periodic tasks.

File: awx/main/dispatch/pool.py [hunk 2]
```diff
@@ -365,6 +367,28 @@ def cleanup(self):
...
+                                current_task['uuid']
+                            ]['started'] = datetime.datetime.utcnow()
+                        age = (
+                            datetime.datetime.utcnow() - current_task['started']
+                        ).total_seconds()
+                        w.managed_tasks[current_task['uuid']]['age'] = age
+                        if age > (60 * 5):
+                            logger.error(
+                                f'run_task_manager has held the advisory lock for >5m, sending SIGTERM to {w.pid}'
+                            )  # noqa
```
Comment (AlanCoding) [inline awx/main/dispatch/pool.py:388]: I cannot figure out why you need the noqa here

**Review by AlanCoding — APPROVED** [review]
The main thing I wanted to see was clear logging for this event, and I can't think of any way I could improve on what you have, so it's all 👍 from me

File: awx/main/dispatch/pool.py [hunk 3]
```diff
@@ -365,6 +367,28 @@ def cleanup(self):
...
+                # the task manager to never do more work
+                current_task = w.current_task
+                if current_task:
+                    if current_task.get('task', '').endswith('tasks.run_task_manager'):
+                        if 'started' not in current_task:
+                            w.managed_tasks[
+                                current_task['uuid']
+                            ]['started'] = datetime.datetime.utcnow()
+                        age = (
+                            datetime.datetime.utcnow() - current_task['started']
```
Comment (AlanCoding) [inline awx/main/dispatch/pool.py]: seems like `time.time` would also be fine. Is it more trustworthy than datetimes??

File: awx/main/dispatch/pool.py (hunk 2 above)
Comment (ryanpetrello) [reply awx/main/dispatch/pool.py:388]: Probably just habit; my editor was complaining about > 80 chars

File: awx/main/dispatch/pool.py (hunk 3 above)
Comment (ryanpetrello) [reply awx/main/dispatch/pool.py]: Yea, that's a good point, I should just do that.

**Comment by AlanCoding** [comment]
deployed this branch and got confirmation that this case is being hit in real installs

```
2020-02-26 02:06:10,932 ERROR    awx.main.dispatch PID:27850 run_task_manager has held the advisory lock for >5m, sending SIGTERM to 10757
2020-02-26 02:06:10,939 WARNING  awx.main.dispatch PID:10757 received SIGTERM, stopping
```

If this instance did not have this patch, it presumably would have hung forever.

**Review by AlanCoding — APPROVED** [review]
Approving it twice because I like it that much

**Comment by ryanpetrello** [comment]
@AlanCoding I'm not _quite_ ready to merge this yet because I want to catch https://github.com/ansible/awx/issues/5617 in action and determine a root cause instead of just papering over it.

**Review by fosterseth — APPROVED** [review]

**Review by elyezer — APPROVED** [review]
LGTM

**Comment by AlanCoding** [comment]
> I'm not quite ready to merge this yet because I want to catch #5617 in action and determine a root cause instead of just papering over it.

While that's fine, I'd like for that to be time-boxed because this bug has been preventing properly testing other features. So I urge an earlier merge. After all, we now have this code that identifies the hangup. Instead of running with `devel`, we could merge this, branch off that, remove the SIGTERM, and send an email or something.

This PR is a nice general fallback, and my attitude toward it would not change even if the root cause is identified.

**Comment by ryanpetrello** [comment]
> While that's fine, I'd like for that to be time-boxed because this bug has been preventing properly testing other features. So I urge an earlier merge.

Okay, I'm fine with that.  I'll deploy and try to reproduce the hang from 9.2.0.

**Comment by AlanCoding** [comment]
regate

**Pull request merged**